web/source/text/processor.ts

```diff
@@ -531,11 +527,19 @@ namespace com.keyman.text {
             DOMEventHandlers.states.changed = true;
           }
         }
+
+        // -- All keystroke (and 'alternate') processing is now complete.  Time to finalize everything! --
         
         // Notify the ModelManager of new input - it's predictive text time!
         ruleBehavior.transcription.alternates = alternates;
         keyman.modelManager.predict(ruleBehavior.transcription);
 
+        // KMEA and KMEI (embedded mode) use direct insertion of the character string
+        if(keyman.isEmbedded) {
+          // A special embedded callback used to setup direct callbacks to app-native code.
+          keyman['oninserttext'](ruleTransform.deleteLeft, ruleTransform.insert, ruleTransform.deleteRight);
+        }
+
         // Since this method now performs changes for 'default' keystrokes, synthetic 'change' event generation
         // belongs here, rather than only in interface.processKeystroke() as in versions pre-12.
         if(outputTarget.getElement()) {
```

Comment:
Of note:  the very next line has already been calling the `doInputEvent` method, hence why there is no visible replacement for the deleted one.